src/optimizer/stats/stats_storage.cpp

```diff
@@ -0,0 +1,318 @@
+    UNUSED_ATTRIBUTE oid_t column_id) {
+  return nullptr;
+}
+
+/**
+ * CollectStatsForAllTables - This function iterates all databases and
+ * datatables
+ * to collect their stats and store them in the 'stats' table.
+ */
+void StatsStorage::CollectStatsForAllTables() {
+  auto catalog = catalog::Catalog::GetInstance();
+
+  oid_t database_count = catalog->GetDatabaseCount();
+  for (oid_t db_offset = 0; db_offset < database_count; db_offset++) {
+    auto database =
+        catalog::Catalog::GetInstance()->GetDatabaseWithOffset(db_offset);
+    oid_t table_count = database->GetTableCount();
+    for (oid_t table_offset = 0; table_offset < table_count; table_offset++) {
+      auto table = database->GetTable(table_offset);
+      std::unique_ptr<TableStats> table_stats(new TableStats(table));
+      table_stats->CollectColumnStats();
+      AddOrUpdateTableStats(table, table_stats.get());
+    }
+  }
+}
+
+/**
+ * CreateSamplesDatabase - Create a database for storing samples tables.
+ */
+void StatsStorage::CreateSamplesDatabase() {
+  catalog::Catalog::GetInstance()->CreateDatabase(SAMPLES_DB_NAME, nullptr);
+}
+
+/**
+ * AddSamplesTable - Add a samples table into the 'samples_db'.
+ * The table name is generated by concatenating db_id and table_id with '_'.
+ */
+void StatsStorage::AddSamplesTable(
+    storage::DataTable *data_table,
+    std::vector<std::unique_ptr<storage::Tuple>> &sampled_tuples) {
+  auto schema = data_table->GetSchema();
+  auto schema_copy = catalog::Schema::CopySchema(schema);
+
+  auto catalog = catalog::Catalog::GetInstance();
+  auto samples_db = catalog->GetDatabaseWithName(SAMPLES_DB_NAME);
+  auto samples_db_oid = samples_db->GetOid();
+  bool own_schema = true;
+  bool adapt_table = false;
+  bool is_catalog = true;
+
+  std::string samples_table_name = GenerateSamplesTableName(
+      data_table->GetDatabaseOid(), data_table->GetOid());
+
+  storage::DataTable *table = storage::TableFactory::GetDataTable(
+      samples_db_oid, catalog->GetNextOid(), schema_copy, samples_table_name,
+      DEFAULT_TUPLES_PER_TILEGROUP, own_schema, adapt_table, is_catalog);
+  samples_db->AddTable(table, true);
+
+  for (auto &tuple : sampled_tuples) {
+    catalog::InsertTuple(table, std::move(tuple), nullptr);
+  }
+}
+
+/**
+ * GetTupleSamples - Query tuple samples by db_id and table_id.
+ * Implement this function.
+ */
+void StatsStorage::GetTupleSamples(
+    UNUSED_ATTRIBUTE oid_t database_id, UNUSED_ATTRIBUTE oid_t table_id,
+    UNUSED_ATTRIBUTE std::vector<storage::Tuple> &tuple_samples) {}
+
+/**
+ * GetColumnSamples - Query column samples by db_id, table_id and column_id.
+ * TODO: Implement this function.
+ */
+void StatsStorage::GetColumnSamples(
+    UNUSED_ATTRIBUTE oid_t database_id, UNUSED_ATTRIBUTE oid_t table_id,
+    UNUSED_ATTRIBUTE oid_t column_id,
+    UNUSED_ATTRIBUTE std::vector<type::Value> &column_samples) {}
+
+} /* namespace optimizer */
+} /* namespace peloton */
```

Comment:
In general this is well written, however do you feel are there going to be more columns added to the stats table at a later stage? If so, then probably making this code more modular especially functions like GetColumnStatsTuple and InitializeStatsSchema might help. Else this seems fine.